### ISSUE TITLE
Over count FLOPS for Tensorflow2

### ISSUE THREAD

**Issue by harryherold** [It4innovations/Intel-SDE-FLOPS#6]
I am using TF profiler to determine the FLOPS required for my Keras model.
I compared these estimated flops with the result of Intel-SDE-FLOPS and they have a big difference!

I tested this on Cascade Lake:
TF Profiler estimates 5500709827 FLOPS for my model and Intel-SDE-FLOPS determines:

Single prec. FLOPs: 11697183045
Double prec. FLOPs: 1538799
Total instructions executed: 12143559585
Total FMA instructions executed: 360400000

On Kaby Lake, I got similar result but with less FMA instructions.
I looked at the sde_mix_out & sde_dyn_mask_profile computed the flops on my own with following results:
Masked Flops: 5896549376 from sde_dyn_mask_profile
Unmasked Flops: 6016063693 from sde_mix_out

Is it really correct that masked & unmasked can be added?
The single values are looking not so bad.
Could you explain me what are the masked FLOPS?

**Comment by gzitzlsb-it4i** [comment]
Hello,

without further details I can only give you a rough answer.

tl;dr:
TF profiler gives you the FLOPs of one feed-forward run of the model's high level operations. Intel SDE counts FLOPs on instruction level. The mapping between both is not 1:1 and it's expected that TF profiler FLOPs <= Intel SDE FLOPs.

More comprehensive explanation:
The TF profiler is not in TF2 yet and so far can only be used via the compat.v1 API using the old tf.session approach. This is not "official" and there could be some problems, too. To my knowledge, it also only computes the required FLOPs of the model itself (feed-forward), which does not consider the number of predictions finally executed. And not only that: also pre-/post-processing or initialization could be counted or not, depending on the description of the model [1]. Also, the FLOPs used for training/fitting and inference/prediction are different, as training executes back-propagation with whatever optimizer was selected. This adds additional computations that the TF profiler does not consider.
Intel SDE FLOPs are precise by counting the executed instructions. However, it only counts the low level FLOPs that are really executed by the processor.
Mapping the model FLOPs to the final processor instruction FLOPs is complex and has to consider questions like:
- How often is the model traversed in the session?
- What else around the model was executed (e.g. pre-/post-processing, initialization)?
- How are the operations implemented and optimized, and which mathematical libraries are they using?
- Do the operations make use of the full SIMD vector length or do they blend elements (esp. for pre-AVX512) which leads to inefficient use of SIMD vectors?
- Is the full SW stack really using the latest SIMD extension (AVX512 in your case) or are there some libraries in between with fallback to older ones? E.g. different BLAS libraries could be used (openBLAS, Intel MKL, etc.).

I'm not surprised that Intel SDE reports more FLOPs than the TF profiler. Why it's around twice as much depends on your model implementation, environment and execution.

In my opinion:
The TF profiler is more applicable to compute the FLOPs of Tensorflow models to compare them on a higher level. Intel SDE measures what is finally executed and provides you information of the efficiency and computational requirements of (not only) the models under execution, depending on the SW stack used.

Answering your second question:
It is correct to add the masked and unmasked counters as implemented. The unmasked (sde-mix-out.txt) ignores any HW mask instructions. These are only covered in the sde-dyn-mask-profile.txt. If you have AVX512 and later which support HW mask instructions, both have to be added (see [2]).

I hope that clarifies the difference in FLOPs you see.

Best regards,

Georg Zitzlsberger


[1]: https://stackoverflow.com/questions/47387561/how-to-calculate-the-flops-from-tfprof-in-tensorflow/50680858
[2]: https://software.intel.com/en-us/articles/calculating-flop-using-intel-software-development-emulator-intel-sde
